tilavarauspalvelu/api/graphql/types/reservation/serializers/update_serializers.py

```diff
@@ -1,114 +1,88 @@
+            "reservee_address_street",
+            "reservee_address_city",
+            "reservee_address_zip",
+            "reservee_is_unregistered_association",
+            "reservee_type",
+            #
+            # Billing information
+            "billing_first_name",
+            "billing_last_name",
+            "billing_email",
+            "billing_phone",
+            "billing_address_street",
+            "billing_address_city",
+            "billing_address_zip",
+            #
+            # Relations
+            "age_group",
             "home_city",
-            "reservee_id",
+            "purpose",
+            #
+            # Read only
+            "state",
         ]
 
-        metadata_set = reservation_unit.metadata_set
-        required_fields = metadata_set.required_fields.order_by("field_name").all() if metadata_set else []
-
-        reservee_type = data.get("reservee_type", getattr(self.instance, "reservee_type", None))
-        if required_fields and reservee_type == CustomerTypeChoice.INDIVIDUAL:
-            required_fields = metadata_set.required_fields.exclude(field_name__in=non_mandatory_fields_for_person)
-
-        for required_field in required_fields:
-            internal_field_name = required_field.field_name
-            existing_value = getattr(self.instance, internal_field_name, None)
-
-            # If the reservee_is_unregistered_association is True it's not mandatory to give reservee_id
-            # even if in metadata set says so.
-            unregistered_field_name = "reservee_is_unregistered_association"
-            if internal_field_name == "reservee_id" and data.get(
-                unregistered_field_name,
-                getattr(self.instance, unregistered_field_name, None),
-            ):
-                continue
-
-            if not data.get(internal_field_name, existing_value):
-                msg = f"Value for required field {to_camel_case(internal_field_name)} is missing."
-                raise ValidationErrorWithCode(
-                    msg,
-                    ValidationErrorCodes.REQUIRED_FIELD_MISSING,
-                    to_camel_case(internal_field_name),
-                )
-
-    def validate(self, data: dict[str, Any]) -> dict[str, Any]:
-        self.check_can_make_changes()
-        self.check_free_of_charge_reason_is_given(data)
-        new_state = self.check_change_state(data)
-
-        data = super().validate(data)
-        data["state"] = new_state
-
-        reservation_units = self._get_reservation_units(data)
-        for reservation_unit in reservation_units:
-            self.check_metadata_fields(data, reservation_unit)
-
+    def validate(self, data: ReservationUpdateData) -> ReservationUpdateData:
+        self.instance.validator.validate_can_change_reservation()
+        self.instance.validator.validate_free_of_charge_arguments(**data)
+        self.instance.validator.validate_required_metadata_fields(**data)
```

Comment:
Hasn't been done before but I guess it should also be validated that the input doesn't attempt to supply fields **not** included in the metadata set (either as required or optional).

Reply:
I think this would be fine and not break the frontend or add additional work, but I'm not 100% sure so let's tag @joonatank.